src/humancellatlas/data/metadata/helpers/staging_area.py

```diff
@@ -218,56 +218,50 @@ def get_bundle_metadata(self, subgraph_id: str) -> Tuple[str, List[JSON], JSON]:
 @attr.s(frozen=True, kw_only=True, auto_attribs=True)
 class GitHubStagingAreaFactory:
     repo: Repository
-    branch: str
+    ref: str
     path: Tuple[str, ...]
 
     @classmethod
-    def from_owner_and_name(cls,
-                            owner: str,
-                            name: str,
-                            branch: str,
-                            path: Optional[str] = None
-                            ) -> 'GitHubStagingAreaFactory':
+    def create(cls,
+               owner: str,
+               name: str,
+               ref: str,
+               path: Optional[str] = None
+               ) -> 'GitHubStagingAreaFactory':
         """
         :param owner: The owner of the GitHub repository
         :param name: The name of the GitHub repository
-        :param branch: A branch name or commit SHA
+        :param ref: A branch name, tag, or commit SHA
         :param path: The path inside the repository to the base of the staging area
         """
         token = os.environ['GH_API_TOKEN']  # A GitHub personal access token
-        logger.debug('Requesting GitHub repo %s', (owner, name, branch, path))
+        logger.debug('Requesting GitHub repo %s', (owner, name, ref, path))
         github_api = Github(token)
         rate_limit = github_api.get_rate_limit()
         logger.debug('GitHub API rate limit limit=%i, remaining=%i, reset=%s',
                      rate_limit.core.limit, rate_limit.core.remaining, rate_limit.core.reset)
         repo = github_api.get_repo(f'{owner}/{name}')
-        if len(branch) == 40 and set(branch).issubset(set('0123456789abcdef')):
-            branch_sha = branch
-        else:
-            gh_branch = repo.get_branch(branch)
-            branch_sha = gh_branch.commit.sha
-            logger.debug('Branch SHA %s', (branch, branch_sha))
         path = () if path == '' else tuple(path.split('/'))
-        return cls(repo=repo, branch=branch_sha, path=path)
+        return cls(repo=repo, ref=ref, path=path)
 
     @classmethod
     def from_url(cls, url: str) -> 'GitHubStagingAreaFactory':
         """
         :param url: A URL of a staging area in a GitHub repository with syntax
-                    'https://github.com/<OWNER>/<NAME>/tree/<BRANCH>[/<PATH>]'.
+                    'https://github.com/<OWNER>/<NAME>/tree/<REF>[/<PATH>]'.
         """
         parsed_url = furl(url)
         require(parsed_url.scheme == 'https', url)
         require(parsed_url.host == 'github.com', url)
         require(len(parsed_url.path.segments) > 3, url)
         require(parsed_url.path.segments[2] == 'tree', url)
         owner, name = parsed_url.path.segments[0:2]
-        branch = parsed_url.path.segments[3]
+        ref = parsed_url.path.segments[3]
```

Comment:
This would break if the URL pointed at a branch whose name had slashes in it, I think.

We should address this by

1) Documenting the restrictions
2) Adding support for a workaround (URL-encode slashes in the branch name) and documenting it
3) Making sure that the code fails in obvious ways if either the ref or the path is not found, or the path doesn't point at a staging area.

Note that a heuristic would be inherently ambiguous. One could set up a repository such that all interpretations of 

`/foo/bar/dir/file`

namely, 

- branch `foo` path `/bar/dir/file`
- branch `foo/bar` path `/dir/file`
- branch `foo/bar/dir` path `/file`
- branch `foo/bar/dir/file` path `/`

would be valid. That's why I think it is not worth exploring a heuristic.

Reply:
1. Will do
2. Everything currently works if `GitHubStagingAreaFactory.create` is given a URL-encoded `ref` value, meaning a branch such as `release-data-24/03/2021` could be specified as `https://github.com/HumanCellAtlas/schema-test-data/tree/release-data-24%2F03%2F2021/tests` when given to `GitHubStagingAreaFactory.from_url`. 
3. I'll see about improving the error messages. Current errors listed below.

If branch name is invalid:
```
github.GithubException.GithubException: 404 {"message": "No commit found for the ref release-data-24/03/2021_foo", "documentation_url": "https://docs.github.com/v3/repos/contents/"}
```

If path is invalid:
```
github.GithubException.UnknownObjectException: 404 {"message": "Not Found", "documentation_url": "https://docs.github.com/rest/reference/repos#get-repository-content"}
```

If path is not to a staging area:
```
humancellatlas.data.metadata.helpers.exception.RequirementError: ['src', 'tests']
```